### DISCUSSION TITLE
cbm_open on peripheral 2 (RS-232) on VIC-20

### DISCUSSION THREAD

**Discussion by sblendorio** [cc65/cc65#1380] [General]
This discussion was previously opened as issue #1375, now closed.

I'm trying to render in CC65 this simple BASIC loop:

    10 open 5,2,3,chr$(6): rem chr$(6) = 300 baud
    20 get#1,a$ : print a$; : get a$ : print #1,a$;
    30 goto 20

But it happens that **cbm_open** fails. This code...

    int code;
    char *str = "x";
    str[0] = 6;
    code = cbm_open(5,2,3,str);

...returns an error (code != 0), so that next commands like **cbm_read** get stuck.

The target is **VIC-20 unexpanded**. This is the repo: https://github.com/sblendorio/victerm300/tree/translate_c

Here is a C program that tries (and fails) to implement that:
https://github.com/sblendorio/victerm300/blob/translate_c/victerm.c#L63

How can I open RS232 peripheral successfully?

I also tried updating the cfg file with this:

        MAIN:     file = %O, define = yes, start = $100D, size = $0BF3 - __STACKSIZE__;

(it was $0DF3)
The result is the same, the program hangs up after calling **cbm_read**.

Thanks in advance

**Comment by nippur72** [comment]
if I understood correctly, the issue is caused by the RS232 buffer being created at the end of memory where the cc65 software stack resides.

If so, a possible fix would be lowering the top memory pointers by 512 just before calling `cbm_open()`. The pointers should be $0037/$0038 and $0283/$0284.

**Reply by sblendorio** [reply]
Solved. I used a custom version of **cbm_read(),** using **cbm_k_getin()** instead of **cbm_k_basin()**, checking accumulator **__A__** instead of checking **cbm_k_readst()**:

    int __fastcall__ cbmread (unsigned char lfn, void* buffer, unsigned int size)
    {
        static unsigned int bytesread;
        static unsigned char tmp;
        unsigned char oserror;
        if (oserror = cbm_k_chkin(lfn)) return -1;
        bytesread = 0;
        while (bytesread<size && !cbm_k_readst()) {
            tmp = cbm_k_getin();
            if (__A__ == 0) break;
            ((unsigned char*)buffer)[bytesread++] = tmp;
        }
        cbm_k_clrch();
        return bytesread;
    }

**Comment by oliverschmidt** [comment]
The return value of `cbm_k_getin()`__*IS*__ the accumulator (https://github.com/cc65/cc65/blob/master/libsrc/cbm/c_getin.s). There's just no point in accessing the accumulator via `__A__` right after returning from `cbm_k_getin()`.

**Reply by nippur72** [reply]
I think he's doing this already as he's not checking against the `0` value anymore. This because it was not possible to receive the ASCII NULL character that some BBS do transmit in order to inject pauses. 

The code now checks for EOI (end of input) by directly inspecting the RS232 buffer pointers: [victerm.c#L161](https://github.com/sblendorio/victerm300/blob/bc1d5b78afac121d1e40d9e52960a6cd035f577a/victerm.c#L161)
